src/main.rs

```diff
@@ -0,0 +1,192 @@
+async fn main() {
+    let cancel = CancellationToken::new();
+    println!("Starting up Consul External DNS");
+
+    println!("=> Parsing configuration");
+    let config = Config::try_parse().expect("===> failed to parse configuration");
+    println!("===> parsed successfully");
+
+    let dns_provider: Box<dyn DnsProviderTrait> = match config.clone().dns_provider {
+        DnsProvider::Hetzner(config) => Box::new(hetzner_dns::HetznerDns { config }),
+    };
+
+    // Initialize Consul Client
+    println!("=> Creating Consul client");
+    let consul_client = create_consul_client(&config).await;
+    println!("===> created Consul client successfully");
+
+    // Create Consul session
+    println!("=> Creating Consul session");
+    let consul_session = consul_client
+        .create_session(cancel.clone())
+        .await
+        .expect("===> failed to create Consul session");
+    let session_id = consul_session.session_id;
+    println!("===> created Consul session successfully");
+
+    // Acquire Lock
+    println!("=> Acquiring Consul lock");
+    if let Err(e) = consul_client.acquire_lock(session_id).await {
+        eprintln!("===> failed to acquire Consul lock: {}", e);
+        return;
+    }
+    println!("===> acquired Consul lock successfully");
+
+    process_dns_records(consul_client, dns_provider, cancel).await;
+
+    consul_session
+        .join_handle
+        .await
+        .expect("Failed to join Consul session handler task");
+}
+
+async fn create_consul_client(config: &Config) -> ConsulClient {
+    loop {
+        match ConsulClient::new(
+            config.consul_address.clone(),
+            // Can not use datacenter until this PR is merged:
+            // https://github.com/hashicorp/consul/pull/21208
+            None,
+        ) {
+            Ok(client) => return client,
+            Err(e) => {
+                eprintln!("===> failed to create Consul client: {}", e);
+                sleep(Duration::from_millis(100)).await;
+            }
+        }
+    }
+}
+
+async fn process_dns_records(
+    consul_client: ConsulClient,
+    dns_provider: Box<dyn DnsProviderTrait>,
+    cancel_token: CancellationToken,
+) {
+    let mut consul_dns_index: Option<String> = None;
+
+    loop {
+        let mut updated_dns_records: HashMap<String, DnsRecord> = HashMap::new();
+
+        // Fetch current DNS records from Consul store
+        println!("=> Fetching DNS records from Consul store");
+        let current_consul_dns_records = match consul_client.fetch_all_dns_records().await {
+            Ok(records) => records,
+            Err(e) => {
+                eprintln!("===> failed to fetch Consul DNS records: {}", e);
+                return;
+            }
+        };
+        println!(
+            "===> fetched Consul DNS records successfully with total records: {}",
+            current_consul_dns_records.len()
+        );
+
+        updated_dns_records.extend(current_consul_dns_records.clone());
```

Comment:
`updated_dns_records` will always be empty before this line so this could instead be `let mut updated_dns_records = current_consul_dns_records.clone();`.

Reply:
fixed